### PR TITLE
HttpApiLogout apiGet/apiPost, bugfixes, gzip, osrm-next preparation

### PR DESCRIPTION
**Pull request by RZR-UA**
**Main:**

All future API get/post requests must be replaced to apiGet() / apiPost()
Forget axios, forget direct fetch. See HttpApiLogout.js for more details.
All existing axios/get/post/fetch calls are already changed and tested.
"302" redirects from API are processed as remote-logout (display loginForm).

**Performance:**

- slow NaN \*\*\*NaN\*\*\* was changed to quickNaNfix() everywhere

**TracksManager bugfixes and improves:**

- getTrackWithAnalysis/getLocalTrackAnalysis able to continue even when get-analysis got fail
- updateRouteBetweenPoints/updateRoute data separated from result (dangerous code)
- temporary bugfix/workaround for localStorage holes (localTrack_0 localTrack_2)

**Other bugfixes include:**

- list-no-keys warnings in formatRouteInfo, RouteSettingsDialog
- "email" cookie did not set after login; SameSite warning fixed

**Other changes:**

- Gzip-Base64 module added (for future localStorage compression)
- globalNavigate() added to call React-route-navigate everywhere
- set Clear ("cross") icon for startPoint and endPoint inputs
- Utils.js: fetchUtil, fetchUtilLoad are removed
- routeShowPoints with GeoJSON filter added
- OSRM-routing preparation

### CODE REVIEW

File: map/src/login/HttpApiLogout.js [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+/*
+    Prepare string with NaN(s) before JSON.parse()
+
+    1) "ele":NaN is converted to "ele":NAN_MARKER
+    2) all others NaN are converted to null
+
+    NaN isn't supported by JSON standard
+*/
+export function quickNaNfix(bad) {
```
Comment (vshcherb) [inline map/src/login/HttpApiLogout.js:243]: This function probably should be in another file not related to logout

File: map/src/util/Utils.js [hunk 2]
```diff
@@ -1,52 +1,53 @@
...
+//     const response = await fetchData()
+//         .then(function (result) {
+//             return result;
+//         });
 
-    if (response.redirected) {
-        setProgressVisible(false);
-        window.location.href = response.url;
-    }
+//     if (response.redirected) {
```
Comment (vshcherb) [inline map/src/util/Utils.js:37]: Delete code as it's not needed

**Comment by RZR-UA** [comment]
2 reviews fixed: commented functions were removed, quickNaNfix was moved to Utils, refs to it were changed

File: map/src/login/HttpApiLogout.js [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    Options:
+
+        default { redirect: 'manual' } block redirect + do logout
+        use { redirect: 'error' } to block redirects + block logout
+        use { redirect: 'follow' } to allow redirects + block logout
+
+        any other options {} are passed to fetch() - { method: 'POST' } as example
+*/
+
+export async function apiGet(url, options = null) {
```
Comment (alisa911) [inline map/src/login/HttpApiLogout.js:85]: apiGet, apiPost move to a separate util file ( mb module)

File: map/src/context/RoutingManager.js [hunk 4]
```diff
@@ -76,11 +76,144 @@ function addSegmentToRouting(start, end, oldPoint, tempPolyline, segments) {
...
+function filterMode(data) {
+    return Object.fromEntries(Object.entries(data).filter(([key]) => !key.includes('rescuetrack')));
+}
+
+// routeProviders
+// called by useState()
+function initRouteProviders() {
+    // fallback default
+    const type = 'osmand';
+    const profile = 'Car';
```
[the remark this answers is not in the document]
Comment (RZR-UA) [reply map/src/context/RoutingManager.js:109]: later (function will use in osrm-next)

File: map/src/context/RoutingManager.js [hunk 5]
```diff
@@ -76,11 +76,144 @@ function addSegmentToRouting(start, end, oldPoint, tempPolyline, segments) {
...
+    const name = 'OsmAnd Advanced Router';
+    const url = `${process.env.REACT_APP_ROUTING_API_SITE}/routing/route`;
+
+    return {
+        // updated via useState()
+        name, // current provider name
```
[the remark this answers is not in the document]
Comment (RZR-UA) [reply map/src/context/RoutingManager.js:115]: dump comments are deleted, routeProviders are commented in better way

File: map/src/login/HttpApiLogout.js [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    Read more details below.
+*/
+
+const loginLogoutForm = '/map/loginForm#logout';
+
+// global useNavigate()
+export let globalNavigate;
+
+// component to include inside <BrowserRoute>
+export const HttpApiLogout = () => {
```
[the remark this answers is not in the document]
Comment (RZR-UA) [reply map/src/login/HttpApiLogout.js:47]: HttpApiLogout was splitted by 2 parts

File: map/src/login/HttpApiLogout.js (hunk 3 above)
Comment (RZR-UA) [reply map/src/login/HttpApiLogout.js:85]: HttpApiLogout was splitted by 2 parts

**Comment by alisa911** [comment]
https://github.com/osmandapp/OsmAnd-Issues/issues/2053

**Comment by alisa911** [comment]
- [ ] Showing errors on UI when changing mail or delete account does not work (setEmailError)
- [ ] Old email in login form after delete account or change email (need empty when delete and new email when change)
- [ ] No need to add obvious information to the logs (POST https://test.osmand.net/mapapi/auth/login 400  and fetch-http-error /mapapi/auth/login, POST https://test.osmand.net/gpx/get-analysis 502 and fetch-http-error /gpx/get-analysis)
- [ ] Button Upload GPX to route does not work correctly - open track in local tracks (not route) and does not check for duplicate names
- [ ] Url is not cleared after delete start and end (https://test.osmand.net/map/?end=50.033989%2C19.903450&mode=bicycle#12/50.0233/19.9862)

File: map/src/App.js [hunk 7]
```diff
@@ -22,4 +22,11 @@ const App = () => {
   );
 };
 
+export let globalNavigate = () => null;
```
Comment (alisa911) [inline map/src/App.js:25]: Inner component should be moved before the main component.

**Comment by RZR-UA** [comment]
Url is not cleared after delete start and end (https://test.osmand.net/map/?end=50.033989%2C19.903450&mode=bicycle#12/50.0233/19.9862)

This is old bug, not related to this PR, so we will fix it later.

**Comment by RZR-UA** [comment]
The button "Upload GPX to route" (renamed from "SELECT TRACK") initially did not act as "Route Mode Track", so I hide this button temporarly, until the Route section will be improved.

**Comment by RZR-UA** [comment]
"Upload GPX to route" - has found and fixed

**Comment by alisa911** [comment]
https://github.com/osmandapp/OsmAnd-Issues/issues/2078

**Pull request merged**